src/kv/kvtypes.h

```diff
@@ -116,6 +103,50 @@ namespace kv
     virtual crypto::Sha256Hash get_root() = 0;
   };
 
+  class Consensus
+  {
+  public:
+    struct NodeConf
+    {
+      NodeId node_id;
+      std::string host_name;
+      std::string port;
+    };
+    virtual ~Consensus() {}
+    virtual bool replicate(
+      const std::vector<std::tuple<Version, std::vector<uint8_t>, bool>>&
+        entries) = 0;
+    virtual Term get_term() = 0; // TODO(#api): this ought to have a more
+                                 // abstract name than Term
+
+    virtual Term get_term(Version version) = 0;
+    virtual Version get_commit_idx() = 0;
+
+    virtual NodeId leader() = 0;
```

Comment:
If we're making the concept of a Leader common to all Consensusesuseises, I think we should implement the shared logic here rather than duplicating/stubbing it elsewhere?

ie - given `leader()` and `id()`, I think we can implement `is_leader()` here as `return leader() == id()` (and similarly `is_follower()`, which should probably directly follow `is_leader()`)? And _maybe_ we could also make `leader` and `local_id` fields of this object and implement `leader()` and `id()` for everyone? And implement the 'obvious' bits of `force_become_leader`?

I think this is neater; the methods remain `virtual` for any implementor who disagrees/wants to extend, but the implementations we _currently_ have take less code.

Reply:
Thanks, can you add this as a next step on the issue #300 so I can take it into consideration? The common interface will need more refactoring in general, trying to keep this PR as small as possible for now.

Reply:
[Done](https://github.com/microsoft/CCF/issues/300#issuecomment-521213035).